airsonic-main/src/main/java/org/airsonic/player/service/search/IndexManager.java

```diff
@@ -1,169 +1,339 @@
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.regex.Pattern;
+
+import static org.springframework.util.ObjectUtils.isEmpty;
+
+/**
+ * Function class that is strongly linked to the lucene index implementation.
+ * Legacy has an implementation in SearchService.
+ * 
+ * If the index CRUD and search functionality are in the same class,
+ * there is often a dependency conflict on the class used.
+ * Although the interface of SearchService is left to maintain the legacy implementation,
+ * it is desirable that methods of index operations other than search essentially use this class directly.
+ */
+@Component
+public class IndexManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(IndexManager.class);
+
+    /**
+     * Schema version of Airsonic index.
+     * It may be incremented in the following cases:
+     * 
+     *  - Incompatible update case in Lucene index implementation
+     *  - When schema definition is changed due to modification of AnalyzerFactory,
+     *    DocumentFactory or the class that they use.
+     *
+     */
+    private static final int INDEX_VERSION = 16;
+
+    /**
+     * Literal name of index top directory.
+     */
+    private static final String INDEX_ROOT_DIR_NAME = "index";
+
+    /**
+     * File supplier for index directory.
+     */
+    private Supplier<File> indexDirectory = () ->
+        new File(SettingsService.getAirsonicHome(), INDEX_ROOT_DIR_NAME);
+
+    /**
+     * File supplier for index version file.
+     * Index version file represents the version of index directory that has already been created.
+     */
+    private Supplier<File> indexVersionFile = () ->
+        new File(indexDirectory.get(), "airsonic".concat(Integer.toString(INDEX_VERSION).concat(".index")));
+
+    /**
+     * Returns the directory of the specified index
+     */
+    private Function<IndexType, File> getIndexDirectory = (indexType) ->
+        new File(indexDirectory.get(), indexType.toString().toLowerCase());
+
+    @Autowired
+    private AnalyzerFactory analyzerFactory;
+
+    @Autowired
+    private DocumentFactory documentFactory;
+
+    private Map<IndexType, SearcherManager> searchers = new HashMap<>();
+
+    private Map<IndexType, IndexWriter> writers = new HashMap<>();
+
+    public void index(Album album) {
+        Term primarykey = documentFactory.createPrimarykey(album);
+        Document document = documentFactory.createAlbumId3Document(album);
+        try {
+            writers.get(IndexType.ALBUM_ID3).updateDocument(primarykey, document);
+        } catch (Exception x) {
+            LOG.error("Failed to create search index for " + album, x);
+        }
+    }
+
+    public void index(Artist artist, MusicFolder musicFolder) {
+        Term primarykey = documentFactory.createPrimarykey(artist);
```

Comment:
Mysterious duplicate checks exist at search result creation because legacy implementations have not paid attention to the uniqueness of index documents.
Currently, documents with the same id do not exist in the search index, so the search result will not contain duplicate ids.

Unlike in the past, this point is guaranteed, so there is no need to check for duplicates after a search.
(There is no problem if you leave it, but if you are concerned that it may cause misunderstanding, it is better to delete it.)